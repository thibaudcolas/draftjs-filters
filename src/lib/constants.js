// @flow
export const ATOMIC = "atomic"
export const UNSTYLED = "unstyled"
export const UNORDERED_LIST_ITEM = "unordered-list-item"
export const ORDERED_LIST_ITEM = "ordered-list-item"

export const IMAGE = "IMAGE"
