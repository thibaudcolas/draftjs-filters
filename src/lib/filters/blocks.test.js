import { EditorState, convertFromRaw, convertToRaw } from "draft-js"

import { UNSTYLED } from "../constants"
import {
  removeInvalidDepthBlocks,
  preserveBlockByText,
  limitBlockDepth,
  filterBlockTypes,
} from "./blocks"

describe("blocks", () => {
  describe("#removeInvalidDepthBlocks", () => {
    it("normalises depth to a given number", () => {
      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "0",
            type: "ordered-list-item",
            depth: 0,
          },
          {
            key: "b",
            text: "1",
            type: "unstyled",
            depth: 1,
          },
          {
            key: "c",
            text: "2",
            type: "unordered-list-item",
            depth: 2,
          },
        ],
      })

      expect(
        removeInvalidDepthBlocks(content)
          .getBlockMap()
          .map((block) => block.getDepth())
          .toJS(),
      ).toEqual({ a: 0, c: 2 })
    })

    it("no normalisation = no change", () => {
      const content = EditorState.createEmpty().getCurrentContent()
      expect(removeInvalidDepthBlocks(content)).toBe(content)
    })
  })

  describe("#limitBlockDepth", () => {
    it("normalises depth to a given number", () => {
      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "0",
            type: "unordered-list-item",
            depth: 0,
          },
          {
            key: "b",
            text: "1",
            type: "unordered-list-item",
            depth: 1,
          },
          {
            key: "c",
            text: "2",
            type: "unordered-list-item",
            depth: 2,
          },
        ],
      })
      expect(
        limitBlockDepth(1, content)
          .getBlockMap()
          .map((block) => block.getDepth())
          .toJS(),
      ).toEqual({ a: 0, b: 1, c: 1 })
    })

    it("no normalisation = no change", () => {
      const content = EditorState.createEmpty().getCurrentContent()
      expect(limitBlockDepth(1, content)).toBe(content)
    })
  })

  describe("#preserveBlockByText", () => {
    it("works", () => {
      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "test",
            type: "unstyled",
          },
          {
            key: "b",
            text: "Â· Bullet 0",
            type: "unstyled",
          },
          {
            key: "c",
            text: "o\tBullet 1",
            type: "unstyled",
          },
          {
            key: "d",
            text: "â—¾Bullet 2",
            type: "unstyled",
          },
          {
            key: "e",
            text: "1. Numbered 1",
            type: "unstyled",
          },
          {
            key: "f",
            text: "a. Numbered 2",
            type: "unstyled",
          },
        ],
      })
      expect(
        convertToRaw(
          preserveBlockByText(
            // WIP. Some of those replacements could be dangerous.
            [
              // https://regexper.com/#%5E(%C2%B7%20%7C%E2%80%A2%5Ct%7C%E2%80%A2%7C%F0%9F%93%B7%20%7C%5Ct%7C%20%5Ct)
              {
                type: "unordered-list-item",
                test: "^(Â· |â€¢\t|â€¢|ðŸ“· |\t| \t)",
                depth: 0,
              },
              { type: "unordered-list-item", test: "^(â—¦|o |o\t)", depth: 1 },
              { type: "unordered-list-item", test: "^(Â§ |ï‚§\t|â—¾)", depth: 2 },
              // https://regexper.com/#%5E(%5Cd%2B%5C.%5B%20%5Ct%5D)
              { type: "ordered-list-item", test: "^\\d+\\.[ \t]", depth: 0 },
              {
                type: "ordered-list-item",
                // Roman numerals from I to XXX.
                test: "^x{0,2}(i|ii|iii|iv|v|vi|vii|viii|ix|x)\\.[ \t]",
                depth: 2,
              },
              {
                type: "ordered-list-item",
                // There is a clash between this and the i. roman numeral. Those tests are executed in order though.
                test: "^[a-z]\\.[ \t]",
                depth: 1,
              },
            ],
            content,
          ),
        ),
      ).toMatchSnapshot()
    })

    it("no normalisation = no change", () => {
      const content = EditorState.createEmpty().getCurrentContent()
      expect(preserveBlockByText([], content)).toBe(content)
    })
  })

  describe("#filterBlockTypes", () => {
    it("works", () => {
      const content = convertFromRaw({
        entityMap: {},
        blocks: [
          {
            key: "a",
            text: "P",
          },
          {
            key: "b",
            text: "UL",
            type: "unordered-list-item",
          },
          {
            key: "c",
            text: "H1",
            type: "header-one",
          },
        ],
      })

      expect(
        filterBlockTypes(["unordered-list-item"], content)
          .getBlockMap()
          .map((block) => block.getType())
          .toJS(),
      ).toEqual({
        a: "unstyled",
        b: "unordered-list-item",
        c: "unstyled",
      })
    })

    it("no normalisation = no change", () => {
      const content = EditorState.createEmpty().getCurrentContent()
      expect(filterBlockTypes([UNSTYLED], content)).toBe(content)
    })
  })
})
