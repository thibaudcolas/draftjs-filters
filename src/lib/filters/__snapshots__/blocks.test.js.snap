// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`blocks #preserveBlockByText works 1`] = `
Object {
  "blocks": Array [
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "a",
      "text": "test",
      "type": "unstyled",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "b",
      "text": "· Bullet 0",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "c",
      "text": "o	Bullet 1",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 2,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "d",
      "text": "◾Bullet 2",
      "type": "unordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 0,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "e",
      "text": "1. Numbered 1",
      "type": "ordered-list-item",
    },
    Object {
      "data": Object {},
      "depth": 1,
      "entityRanges": Array [],
      "inlineStyleRanges": Array [],
      "key": "f",
      "text": "a. Numbered 2",
      "type": "ordered-list-item",
    },
  ],
  "entityMap": Object {},
}
`;
